### PR TITLE
Added configurable bound to the ingest metric timespace

### PR DESCRIPTION
**Pull request by rajeshdasi**
# Resolves

https://jira.rax.io/browse/CERES-383

# What

This PR adds bounds to the ingest metric timestamps. In case no value is provided default values are set.

### CODE REVIEW

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java [hunk 1]
```diff
@@ -110,4 +110,8 @@
   @NotNull Integer tagValueLimit = 50;
 
   List<String> excludedTagKeys;
+
+  String ingestStartTime;
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/config/AppProperties.java:114]: Take this as Duration object instead of String

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java [hunk 2]
```diff
@@ -33,49 +33,49 @@
...
 public class AppProperties {
+
   @NotNull
   @DurationUnit(ChronoUnit.SECONDS)
   Duration rawTtl = Duration.ofHours(6);
 
   /**
-   * When initially creating the Cassandra data table schemas,
-   * this value will be used for the expired data garbage collection.
+   * When initially creating the Cassandra data table schemas, this value will be used for the
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/config/AppProperties.java:42]: Remove this unnecessary changes in comments.

File: src/main/java/com/rackspace/ceres/app/web/WriteController.java [hunk 3]
```diff
@@ -21,7 +21,13 @@
 import com.rackspace.ceres.app.model.PutResponse;
 import com.rackspace.ceres.app.model.TagFilter;
 import com.rackspace.ceres.app.services.DataWriteService;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalUnit;
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/web/WriteController.java:27]: Unused imports added I guess.

**Review by rohitsngh27 — CHANGES_REQUESTED** [review]
Some changes need to be done related to Duration

File: src/main/java/com/rackspace/ceres/app/web/WriteController.java [hunk 4]
```diff
@@ -21,7 +21,13 @@
 import com.rackspace.ceres.app.model.PutResponse;
 import com.rackspace.ceres.app.model.TagFilter;
 import com.rackspace.ceres.app.services.DataWriteService;
+
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/web/WriteController.java:24]: You can remove this file from the diff?

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java [hunk 5]
```diff
@@ -33,49 +33,49 @@
...
    */
   @Min(20)
   int rawCompactionWindows = 30;
 
   /**
-   * The width of time slots used for partitioning data, which can reduce the number
-   * of Cassandra files that need to be scanned. It is best to set this to the same
-   * as <code>ceres.downsample.timeSlotWidth</code> or greater to ensure downsampling
-   * queries only need to consult one partition.
+   * The width of time slots used for partitioning data, which can reduce the number of Cassandra
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/config/AppProperties.java:56]: These edits seem unrelated to the topic of this PR? Shouldn't they be in a separate PR?

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+  public IngestBoundsValidator(AppProperties appProperties) {
+    this.appProperties = appProperties;
+  }
+
+  @Override
+  public boolean isValid(Instant value, ConstraintValidatorContext context) {
+    log.info("Field Value - {}", value);
+    value = value.atZone(ZoneId.of("UTC")).toInstant();
+    return !(value.isBefore(getStartTime()) || value.isAfter(getEndTime()));
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:28]: The logic is a bit backwards, I would prefer !value.isBefore(getStartTime() && !value.isAfter(getStartTime())

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * This method calculates the start ingested metrics time, in case no value is provided the
+   * default value is set to 1 week in the past.
+   *
+   * @return - {@code Instant}
+   */
+  private Instant getStartTime() {
+    int startParams;
+    if (appProperties.getIngestStartTime() == null || appProperties.getIngestStartTime()
+        .isEmpty()) {
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:40]: You are calling appProperties.getIngestStartTime() 3 times, why not save it in a variable?

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      startParams = 7;
+    } else {
+      String extractStartTime = appProperties.getIngestStartTime().replaceAll("[^0-9]", "");
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:43]: Looks like you are are erasing the numbers. I must be missing something here...

**Review by matsboman — COMMENTED** [review]
The validation looks a bit complex. Isn't there a way to make the code simpler?

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  /**
+   * This method calculates the start ingested metrics time, in case no value is provided the
+   * default value is set to 1 week in the past.
+   *
+   * @return - {@code Instant}
+   */
+  private Instant getStartTime() {
+    int startParams;
```
[the remark this answers is not in the document]
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:38]: In case if we use "Duration.between" we need to validate the result again instead as we are holding the value in Instant we can rely on the inbuild function isAfter & isBefore.

File: src/main/java/com/rackspace/ceres/app/web/WriteController.java (hunk 3 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/web/WriteController.java:27]: Fixed it

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java (hunk 5 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/config/AppProperties.java:56]: Fixed it.

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java (hunk 6 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:28]: Fixed it

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java (hunk 7 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:40]: Fixed it

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java (hunk 8 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:43]: Fixed it

File: src/main/java/com/rackspace/ceres/app/web/WriteController.java (hunk 4 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/web/WriteController.java]: By mistake, I had committed my changes with unused imports. I'll make sure that this is not repeated.

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java (hunk 1 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/config/AppProperties.java]: Fixed it

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java (hunk 2 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/config/AppProperties.java]: Fixed it

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java [hunk 10]
```diff
@@ -110,4 +110,8 @@
   @NotNull Integer tagValueLimit = 50;
 
   List<String> excludedTagKeys;
-}
+
+  Duration ingestStartTime;
+
+  Duration ingestEndTime;
+}
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/config/AppProperties.java:117]: Missing line at EOF

**Review by matsboman — COMMENTED** [review]
LGTM, just one missing line at EOF

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+
+@Log4j2
```
Comment (shabd67) [inline src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:34]: are you using any logging here?

**Review by shabd67 — APPROVED** [review]
looks good to me. let's wait for others.

**Review by matsboman — DISMISSED** [review]
LGTM

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java (hunk 11 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java]: FIxed it

File: src/main/java/com/rackspace/ceres/app/config/AppProperties.java (hunk 10 above)
Comment (rajeshdasi) [reply src/main/java/com/rackspace/ceres/app/config/AppProperties.java]: Fixed it

File: src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import javax.validation.ConstraintValidatorContext;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+
+@Component
+@Profile("ingest")
+public class IngestBoundsValidator implements ConstraintValidator<IngestBounds, Instant> {
+
+  private final AppProperties appProperties;
+  private Duration startDuration = Duration.ofDays(7);
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/ceres/app/validation/IngestBoundsValidator.java:38]: Why did you hardcode this value here? This violates the process of making ingestStartTime and ingestEndTime as configurable.

**Review by rohitsngh27 — APPROVED** [review]
Looks much cleaner to me now. Let's wait for others to review

File: src/main/resources/application-dev.yml [hunk 13]
```diff
@@ -14,7 +14,9 @@ ceres:
     partitions-to-process: 0-3
   tag-filter: truncate
   tag-value-limit: 50
-  excluded-tag-keys: ["ip_address"]
+  excluded-tag-keys: [ "ip_address" ]
```
Comment (shabd67) [inline src/main/resources/application-dev.yml:17]: why to change excluded-tag-keys value? I see an extra space there.

**Review by shabd67 — APPROVED** [review]

**Review by matsboman — APPROVED** [review]

**Comment by rajeshdasi** [comment]
@harageth For future reference.
